1. If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000.


In [7]:
natural_1000 = list(range(1,1000))
mult_3 = []
mult_5 = []
for n in natural_1000:
    if n % 3 == 0:
        mult_3.append(n)
for n in natural_1000:
    if n % 5 == 0:
        mult_5.append(n)
sum_3 = sum(mult_3)
sum_5 = sum(mult_5)
total_sum = sum_3 + sum_5
total_sum

266333

2. Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:

1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...

By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.


In [19]:
fibo = [1,2]
j = 0
while fibo[-1] < 4000000:
    fibo.append((fibo[j]+fibo[j+1]))
    j += 1
fibo_pairs = []
for n in fibo:
    if n % 2 == 0:
        fibo_pairs.append(n)

sum(fibo_pairs)


4613732

3. The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?


In [15]:
def largest_prime_factor(n):
    i = 2
    while i * i <= n:
        if n % i:
            i += 1
        else:
            n //= i
    return n

print(largest_prime_factor(600851475143))

6857


4. A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.


In [42]:
d3number = list(range(100, 1000))
def reverse(number):
    m = str(number)
    rev = m[::-1]
    return rev

palin = []

for n in d3number:
    for m in d3number:
        p = n*m
        if str(p) == reverse(p):
            palin.append(p)


largest = max(palin)

largest

906609

5. 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?


In [56]:
import math
list_20 = list(range(1,21))
n = 2
factor = list_20[1]
while n < len(list_20):
    mcm = abs(factor*list_20[n])//math.gcd(factor,list_20[n])
    factor = mcm
    n += 1
factor

232792560

6. The sum of the squares of the first ten natural numbers is, 1^2 + 2^2 + ... + 10^2 = 385

The square of the sum of the first ten natural numbers is, (1 + 2 + ... + 10)^2 = 3025

Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is, 3025 - 385 = 2640

Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

In [62]:
n100 = list(range(1,101))
qr = []
numqr = []
for n in n100:
    qr.append(n**2)
for n in n100:
    numqr.append(n)
diff = sum(numqr)**2 - sum(qr)
diff

25164150

7. By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

What is the 10001st prime number?


In [64]:
import math

def is_prime(n):
    if n < 2:
        return True
    for i in range(2,int(math.sqrt(n))+1):
        if n % i == 0:
            return False
    return True

prime = [2,3,5,7,11,13]
n = 14
while len(prime) < 10001:
    if is_prime(n):
        prime.append(n)
        n += 1
    else:
        n += 1
prime[-1]

104743

8. A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,
a^2 + b^2 = c^2

For example, 3^2 + 4^2 = 9 + 16 = 25 = 5^2.

There exists exactly one Pythagorean triplet for which a + b + c = 1000.
Find the product abc.


In [55]:
for a in range(1, 1000):
    for b in range(a+1, 1000):
        c = 1000 - a - b
        if a**2 + b**2 == c**2:
            print(a*b*c)
            break
            

31875000


9. The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

Find the sum of all the primes below two million.


In [73]:
import math

def is_prime(n):
    if n < 2:
        return True
    for i in range(2,int(math.sqrt(n))+1):
        if n % i == 0:
            return False
    return True
prime = [1,2,3,5,7,11,13]
n = 14
while prime[-1] < 2E+06:
    if is_prime(n):
        prime.append(n)
        n += 1
    else:
        n += 1
else:
    n += 1

sum(prime)

142915828926

10. The sequence of triangle numbers is generated by adding the natural numbers. So the 7th triangle number would be 1 + 2 + 3 + 4 + 5 + 6 + 7 = 28. The first ten terms would be:

1, 3, 6, 10, 15, 21, 28, 36, 45, 55, ...

Let us list the factors of the first seven triangle numbers:

     1: 1
     3: 1,3
     6: 1,2,3,6
    10: 1,2,5,10
    15: 1,3,5,15
    21: 1,3,7,21
    28: 1,2,4,7,14,28

We can see that 28 is the first triangle number to have over five divisors.

What is the value of the first triangle number to have over five hundred divisors?


In [54]:
import math

def is_prime(number):
    if number < 2:
        return True
    for i in range(2,int(math.sqrt(number))+1):
        if number % i == 0:
            return False
    return True

def factorize(number):
    '''This fucntions returns a set with all the factors of a given number'''
    if is_prime(number):
        return {1,number}
    else:
        factors = set()
        div = set((range(1,number+1)))
        for n in div:
            if number % n == 0:
                factors.add(n)
            else:
                continue
        return factors


#Defining the initial variables

n = 2000
tn = int((n*(n+1))/2)



def wloop(n,tn,fc):
    '''This function return the th number and its value that its number of factors is over a given number of factors.'''
    while len(factorize(tn)) <= fc:
        n += 1
        tn = int(((n+1)*n)/2)
    return n, tn

n, tn = wloop(n,tn,300)



print(f'{n}th, {tn}, {len(factorize(tn))}')

2079th, 2162160, 320


Considerations: Prime numbers only have two factors. Discarting all those numbers will help the program.

n = 2079 --> 320 factors


n = ; tn = ; --> 500 factors